base/range.jl

```diff
@@ -1263,13 +1263,17 @@ function -(r::LinRange)
     LinRange{typeof(start)}(start, -r.stop, length(r))
 end
 
-
 # promote eltype if at least one container wouldn't change, otherwise join container types.
-el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{T,n}}) where {T,n}   = a
+el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{T,n}}) where {T,n}   = a # we assume a === b
 el_same(::Type{T}, a::Type{<:AbstractArray{T,n}}, b::Type{<:AbstractArray{S,n}}) where {T,S,n} = a
 el_same(::Type{T}, a::Type{<:AbstractArray{S,n}}, b::Type{<:AbstractArray{T,n}}) where {T,S,n} = b
 el_same(::Type, a, b) = promote_typejoin(a, b)
 
+promote_result(::Type{<:AbstractArray}, ::Type{<:AbstractArray}, ::Type{T}, ::Type{S}) where {T,S} = (@inline; promote_type(T,S))
+promote_result(::Type{T}, ::Type{S}, ::Type{Bottom}, ::Type{Bottom}) where {T<:AbstractArray,S<:AbstractArray} = (@inline; promote_typejoin(T,S))
+# If no promote_rule is defined, both directions give Bottom. In that case use typejoin on the eltypes instead and give Array as the container.
+promote_result(::Type{<:AbstractArray{T,n}}, ::Type{<:AbstractArray{S,n}}, ::Type{Bottom}, ::Type{Bottom}) where {T,S,n} = (@inline; Array{promote_type(T,S),n})
```

Comment:
@vtjnash Line 1275 breaks `StaticArrays.jl`'s test on master.
After this PR
```julia
julia> [(@MArray [1;2;3]), (@SArray [1;2;3])]
2-element Vector{Vector{Int64}}:
 [1, 2, 3]
 [1, 2, 3]
```
It was a `Vector{Any}` before this PR.
`for a in [array1, array2]` is general pattern for test. So this change seems a little "break"?

Reply:
Test should not do that, that is an incorrect test. `[a, b]` is a general pattern for changing `a` and `b` to a common type. If you do not want that, you must specify the type of the array you want (usually `Any[...]`)

Reply:
FWIW CategoricalArrays tests also used this pattern. Actually this revealed the lack of `promote_type` methods for `CategoricalArray`s.

This has the potential to disrupt lots of tests, as things like e.g. `[["b", "a"], ["b", missing]]` will suddenly give two `Vector{Union{Missing, String}}` objects while previously the first one was `Vector{String}`. I'm not saying it shouldn't be done though but we have to be ready to notice some breakage.